### PR TITLE
ORC-1298:[C++] Support dedicated ColumnVectorBatch of numeric types

### PR DESCRIPTION
**Pull request by luffy-zh**
### What changes were proposed in this pull request?

This patch implements int8/int16/int32/float buffer and refactors the RLE encoder/decoder, IntegerColumnWriter/Reader, and DoubleColumnWriter/Reader.

### Why are the changes needed?

For the context of [ORC-1298](https://issues.apache.org/jira/projects/ORC/issues/ORC-1298?filter=allopenissues), this change makes user no need to focus on overflow.

### How was this patch tested?

Add the WriterTest.testWriterFixedWidthNumericVectorBatch UT.

### CODE REVIEW

**Comment by wgtmac** [comment]
Thanks @luffy-zh for the contribution!

cc @stiga-huang @coderex2522

File: c++/include/orc/Type.hh [hunk 1]
```diff
@@ -70,10 +70,11 @@ namespace orc {
...
      */
-    virtual ORC_UNIQUE_PTR<ColumnVectorBatch> createRowBatch(uint64_t size,
-                                                             MemoryPool& pool,
-                                                             bool encoded = false
-                                                             ) const = 0;
+    virtual ORC_UNIQUE_PTR<ColumnVectorBatch> createRowBatch(
+      uint64_t size,
+      MemoryPool& pool,
+      bool encoded = false,
+      bool enableFixedWidthNumericVectorBatch = false) const = 0;
```
Comment (dongjoon-hyun) [inline c++/include/orc/Type.hh:77]: @luffy-zh When you add one parameter, please avoid any change to the existing parameters. For example, indentation changes on line 73 to 76.

Comment (dongjoon-hyun) [reply c++/include/orc/Type.hh:77]: In addition, can we keep the original three-parameter method and add new four-parameter method with new parameter additionally?
```
bool enableFixedWidthNumericVectorBatch = false) const = 0;
```

File: c++/src/ColumnWriter.cc [hunk 2]
```diff
@@ -456,126 +456,107 @@ namespace orc {
...
   class IntegerColumnWriter : public ColumnWriter {
   public:
     IntegerColumnWriter(
                         const Type& type,
                         const StreamsFactory& factory,
-                        const WriterOptions& options);
+                        const WriterOptions& options) :
+                          ColumnWriter(type, factory, options),
+                          rleVersion(options.getRleVersion()) {
+      std::unique_ptr<BufferedOutputStream> dataStream =
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:468]: nit: you can still separate template class declaration with definition like before to keep the class declaration compact. See here for reference: https://stackoverflow.com/questions/13351474/c-separate-declaration-definition-for-template-function-in-template-class

File: c++/src/ColumnWriter.cc [hunk 3]
```diff
@@ -456,126 +456,107 @@ namespace orc {
     }
   }
 
+  template <typename BatchType, typename IntType>
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:459]: IntType can be deduced from BatchType, isn't it?

File: c++/src/ColumnWriter.cc [hunk 4]
```diff
@@ -828,16 +810,17 @@ namespace orc {
...
-                          bool isFloatType) :
-                              ColumnWriter(type, factory, options),
-                              isFloat(isFloatType),
-                              buffer(*options.getMemoryPool()) {
+  template<typename ValueType, typename BatchType>
+  DoubleColumnWriter<ValueType, BatchType>::DoubleColumnWriter(
+    const Type& type,
+    const StreamsFactory& factory,
+    const WriterOptions& options,
+    bool isFloatType) :
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:818]: We should throw if isFloatType == false and ValueType is float

File: c++/src/ColumnWriter.cc [hunk 5]
```diff
@@ -3061,15 +3050,36 @@ namespace orc {
...
+                                                             options));
+        }
       case LONG:
-      case SHORT:
         return std::unique_ptr<ColumnWriter>(
-          new IntegerColumnWriter(
-                                  type,
-                                  factory,
-                                  options));
+          new IntegerColumnWriter<LongVectorBatch, int64_t>(
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:3071]: It seems redundant to define the 2nd template parameter

File: c++/src/ColumnWriter.cc [hunk 6]
```diff
@@ -3083,18 +3093,26 @@ namespace orc {
                                   options));
       case DOUBLE:
         return std::unique_ptr<ColumnWriter>(
-          new DoubleColumnWriter(
-                                 type,
-                                 factory,
-                                 options,
-                                 false));
+          new DoubleColumnWriter<double, DoubleVectorBatch>(
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:3096]: ditto

File: c++/src/RLE.hh [hunk 7]
```diff
@@ -59,8 +59,16 @@ namespace orc {
...
-                      const char* notNull);
+                     const char* notNull);
 
+    virtual void add(const int32_t* data, uint64_t numValues,
+                     const char* notNull);
+
+    virtual void add(const int16_t* data, uint64_t numValues,
+                     const char* notNull);
+
+    virtual void add(const int8_t* data, uint64_t numValues,
```
Comment (wgtmac) [inline c++/src/RLE.hh:70]: RLE only supports int64_t, int32_t, and int16_t. Please remove the int8_t overload.

File: c++/src/RLE.hh [hunk 8]
```diff
@@ -126,6 +134,15 @@ namespace orc {
     virtual void next(int64_t* data, uint64_t numValues,
                       const char* notNull) = 0;
 
+    virtual void next(int32_t* data, uint64_t numValues,
+                      const char* notNull) = 0;
+
+    virtual void next(int16_t* data, uint64_t numValues,
+                      const char* notNull) = 0;
+
+    virtual void next(int8_t* data, uint64_t numValues,
```
Comment (wgtmac) [inline c++/src/RLE.hh:143]: ditto

File: c++/src/RLE.cc [hunk 9]
```diff
@@ -71,14 +71,41 @@ namespace orc {
...
-                         const char* notNull) {
+                       const char* notNull) {
     for (uint64_t i = 0; i < numValues; ++i) {
       if (!notNull || notNull[i]) {
         write(data[i]);
       }
     }
   }
 
+  void RleEncoder::add(const int32_t* data, uint64_t numValues,
```
Comment (wgtmac) [inline c++/src/RLE.cc:82]: Why not use a template function to repeat the code? static_cast does not have any runtime penalty.

File: c++/include/orc/Vector.hh [hunk 10]
```diff
@@ -88,28 +89,88 @@ namespace orc {
...
-    void resize(uint64_t capacity);
-    void clear();
-    uint64_t getMemoryUsage();
+    virtual ~IntegerVectorBatch() {
+      // PASS
+    }
+
+    std::string toString() const {
+      std::ostringstream buffer;
+      buffer << "Long vector <" << numElements << " of " << capacity << ">";
```
Comment (wgtmac) [inline c++/include/orc/Vector.hh:106]: It may not be "Long vector" any more

File: c++/include/orc/Vector.hh [hunk 11]
```diff
@@ -88,28 +89,88 @@ namespace orc {
...
+      // PASS
+    }
+
+    virtual ~BaseDoubleVectorBatch() {
+      // PASS
+    }
+
+    std::string toString() const {
+      std::ostringstream buffer;
+      buffer << "Double vector <" << numElements << " of " << capacity << ">";
```
Comment (wgtmac) [inline c++/include/orc/Vector.hh:148]: It may not be "Double vector" any more.

File: c++/test/TestWriter.cc [hunk 12]
```diff
@@ -2044,5 +2048,91 @@ namespace orc {
     testSuppressPresentStream(CompressionKind_SNAPPY);
   }
 
+  TEST_P(WriterTest, testWriteFixedWidthNumericVectorBatch) {
+    MemoryOutputStream memStream(DEFAULT_MEM_STREAM_SIZE);
+    MemoryPool * pool = getDefaultPool();
+    std::unique_ptr<Type> type(Type::buildTypeFromString(
+      "struct<col1:double,col2:float,col3:int,col4:smallint,col5:tinyint>"));
```
Comment (coderex2522) [inline c++/test/TestWriter.cc:2055]: It's better to test the bigint type as well.

File: .gitignore [hunk 13]
```diff
@@ -1,14 +0,0 @@
-build
```
Comment (coderex2522) [inline .gitignore:1]: This file is not associated with this pull request. Whether it is a misoperation?

File: c++/src/ColumnReader.cc [hunk 14]
```diff
@@ -474,31 +465,32 @@ namespace orc {
...
+  void DoubleColumnReader<columnKind, isLittleEndian, ValueType, BatchType>::next(
+      ColumnVectorBatch& rowBatch, uint64_t numValues, char* notNull) {
     ColumnReader::next(rowBatch, numValues, notNull);
     // update the notNull from the parent class
     notNull = rowBatch.hasNulls ? rowBatch.notNull.data() : nullptr;
-    double* outArray = dynamic_cast<DoubleVectorBatch&>(rowBatch).data.data();
+    ValueType* outArray =
+        reinterpret_cast<ValueType*>(dynamic_cast<BatchType&>(rowBatch).data.data());
 
     if (columnKind == FLOAT) {
```
Comment (wgtmac) [inline c++/src/ColumnReader.cc:477]:
```suggestion
    if constexpr (columnKind == FLOAT) {
```

File: c++/src/Timezone.cc [hunk 15]
```diff
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
+#include <iostream>
```
Comment (wgtmac) [inline c++/src/Timezone.cc:27]: Remove it

File: tools/src/FileMetadata.cc [hunk 16]
```diff
@@ -26,7 +26,7 @@
 #include "orc/Exceptions.hh"
 #include "orc/OrcFile.hh"
 
-//#include "Adaptor.hh"
+// #include "Adaptor.hh"
```
Comment (wgtmac) [inline tools/src/FileMetadata.cc:29]: Remove this line

**Review by wgtmac — CHANGES_REQUESTED** [review]

File: c++/src/MemoryPool.cc [hunk 17]
```diff
@@ -180,6 +198,60 @@ namespace orc {
...
+
+  template <>
+  DataBuffer<int16_t>::~DataBuffer() {
+    if (buf) {
+      memoryPool.free(reinterpret_cast<char*>(buf));
+    }
+  }
+
+  template <>
+  void DataBuffer<int16_t>::resize(uint64_t newSize) {
```
Comment (wgtmac) [inline c++/src/MemoryPool.cc:229]: resize of different vectors should also be in the unit test

File: c++/src/ColumnWriter.cc [hunk 18]
```diff
@@ -415,109 +415,97 @@ namespace orc {
...
+        recordPosition();
+      }
+    }
 
-    virtual void flush(std::vector<proto::Stream>& streams) override;
+    virtual void add(ColumnVectorBatch& rowBatch, uint64_t offset, uint64_t numValues,
+                     const char* incomingMask) override {
+      const BatchType* intBatch = dynamic_cast<const BatchType*>(&rowBatch);
+      if (intBatch == nullptr) {
+        throw InvalidArgument("Failed to cast to LongVectorBatch");
```
Comment (wgtmac) [inline c++/src/ColumnWriter.cc:438]:
```suggestion
        throw InvalidArgument("Failed to cast to integer vector batch");
```

**Review by wgtmac — APPROVED** [review]
LGTM +1

File: c++/src/ColumnReader.cc [hunk 19]
```diff
@@ -1696,13 +1690,25 @@ namespace orc {
...
     switch (static_cast<int64_t>(type.getKind())) {
-      case DATE:
-      case INT:
+      case SHORT: {
+        if (useTightNumericVector) {
+          return std::unique_ptr<ColumnReader>(
+              new IntegerColumnReader<ShortVectorBatch>(type, stripe));
+        }
+      }
+      case INT: {
```
Comment (dongjoon-hyun) [inline c++/src/ColumnReader.cc:1702]: This seems to cause `implicit-fallthrough` warning and compilation failure.

Comment (wgtmac) [reply c++/src/ColumnReader.cc:1702]: Should we fix it?

Comment (dongjoon-hyun) [reply c++/src/ColumnReader.cc:1702]: Yes, we need to fix or mitigate this for Apache ORC 1.9.0 because this causes Docker Test failures.

Comment (dongjoon-hyun) [reply c++/src/ColumnReader.cc:1702]: I filed a JIRA and PR for `branch-1.9` first. Will make another PR for main branch..
- https://issues.apache.org/jira/browse/ORC-1453
- https://github.com/apache/orc/pull/1548 (for branch-1.9)

**Pull request merged**